Allow length of readUtf8LineStrict to be limited.

ericedens opened the pull request:
I ran into needing this while using okio to write an HTTP server, and I think it should address issue #153 .

Two API additions:

1. BufferedOutputStream.readUtf8LineStrict accepts a maxByteCount parameter. If the line break isn't found by that limit, an EOFException is thrown.  (Is that the right exception??)
2. BufferedOutputStream.indexOf includes a marker for ending the search. If reached, -1 is returned.

I thought about whether readUtf8Line should get the maxByteCount parameter as well. If added, it would have a contract of questionable utility: return a string at end of line, EOF, or when this count parameter is reached. In any case, if the method is desirable, it should be easy to add.

Additionally, there are indexOf methods that search for ByteStrings. Those aren't covered by this PR either, but should be similarly easy to add if wanted.

Test output: [mvn-clean-verify.txt](https://github.com/square/okio/files/692663/mvn-clean-verify.txt)
Related issue: #153 Optional limits on methods like readLine()

swankjesse commented:
Awesome. I've been dragging my feet thinking about how to expose this and you've just gone and done it. Nice. I'll give it a full review soon!

JakeWharton commented on okio/src/test/java/okio/ReadUtf8LineTest.java:
```diff
@@ -94,6 +94,20 @@
+      source.readUtf8LineStrict(4);
+      fail();
+    } catch (EOFException expected) {
+      assertEquals("\\n not found: size=4 content=6162636465660a…", expected.getMessage());
```
This message looks odd. Size 4 but content is more than 4 bytes.

ericedens replied:
It does look odd, now that you mention it. One option: "\n not found in first n bytes of content=6162636465660a…", where n is replaced with the number of bytes scanned (in this case 4).

cketti replied:
Rather than including encoded bytes in the exception message it'd be nice to have a guarantee (JavaDoc and tests) that the BufferedSource is still in a useable state so users could fetch the information themselves if necessary.

JakeWharton replied:
Yes, that would be a good addition.

ericedens replied:
The byte dump is an existing feature, so I'll let Jesse and Jake weigh in on whether they want to keep that.

The previous implementation still follows: the bytes are copied from the source instead of consumed. Regardless, let's definitely add tests and javadocs to protect that behavior.

swankjesse commented on okio/src/main/java/okio/Buffer.java:
```diff
@@ -1263,18 +1267,23 @@ Segment writableSegment(int minimumCapacity) {
 
     Segment s;
     long offset;
+    if (toIndex > size) toIndex = size;
```
this should probably throw. It’s probably a bug.

ericedens replied:
sgtm

ericedens replied:
After some more thought, I don't think throwing is appropriate. The buffer may not be populated with all of the stream's future data yet, even though the caller still knows that they're willing to read up to n bytes. This happens in RealBufferedSource.indexOf, where additional data may be read from the source if the target byte isn't found yet.

swankjesse commented on okio/src/main/java/okio/Buffer.java:
```diff
@@ -1283,9 +1292,9 @@ Segment writableSegment(int minimumCapacity) {
         // No segments to scan!
         return -1L;
-      } else if (size - fromIndex < fromIndex) {
+      } else if (toIndex - fromIndex < fromIndex) {
```
This should still be size. Otherwise incorrect. This is the offset of the scan, not the desired offset of the scan. It’ll be broken for buffers with multiple segments.

ericedens replied:
yuck, good catch.

swankjesse commented on okio/src/main/java/okio/Buffer.java:
```diff
@@ -1283,9 +1292,9 @@ Segment writableSegment(int minimumCapacity) {
+      } else if (toIndex - fromIndex < fromIndex) {
         // We're scanning in the back half of this buffer. Find the segment starting at the back.
-        offset = size;
+        offset = toIndex;
```
same

ericedens replied:
fixed

swankjesse commented on okio/src/main/java/okio/Buffer.java:
```diff
@@ -1283,9 +1292,9 @@ Segment writableSegment(int minimumCapacity) {
         // We're scanning in the back half of this buffer. Find the segment starting at the back.
-        offset = size;
+        offset = toIndex;
         while (offset > fromIndex) {
```
... and this code will need to just skip segments until `offset <= toIndex`

ericedens replied:
Rolled back to the original version -- we still need to find the segment that contains toIndex, because that's where the scan will start.

swankjesse commented on okio/src/main/java/okio/Buffer.java:
```diff
@@ -1301,9 +1310,10 @@ Segment writableSegment(int minimumCapacity) {
+    while (offset < toIndex) {
       byte[] data = s.data;
-      for (int pos = (int) (s.pos + fromIndex - offset), limit = s.limit; pos < limit; pos++) {
+      long limit = Math.min(s.limit, s.pos + toIndex - offset);
```
dig

swankjesse commented on okio/src/main/java/okio/BufferedSource.java:
```diff
@@ -433,6 +440,13 @@
   long indexOf(byte b, long fromIndex) throws IOException;
 
+  /**
+   * Returns the index of {@code b} if it is found in the range of {@code fromIndex} inclusive
```
nice docs

ericedens replied:
thanks!

swankjesse commented:
I think this PR is great.

Next steps:

 * We need some tests that cover multiple-segment buffers to confirm that those cases work. My guess is that if you have, say, 5 segments and it starts from the end in the 3rd segment, it’s going to get the wrong answer.
 * Review other APIs that need this. (`readUtf8Line()`, `indexOf(ByteString)`, `indexOfElement()`). You’re welcome to do this work if you’d like, or I can take it. You might enjoy it!

ericedens commented:
Thanks for the review! In the next few days I'll update those spots, and take a look at the other methods that could use an upper bound.

ericedens commented:
OK! Here it is. I fixed the items from your comments, and added some multi-segment tests. (Travis had a series of out-of-memory failures from multi-segment ONE_BYTE_AT_A_TIME tests, but those are tuned now.)

For the other APIs, my preference would be to do those in another PR so that the change doesn't get too big. Thoughts?

Thanks for taking a look!

swankjesse commented:
I merged this manually. Really nice PR. Thank you.

swankjesse commented:
Also I made two changes. I renamed `maxRead` to `limit`. I think that’s slightly more consistent with the names in Okio. And I changed the code to accept `limit = 0` which will throw an `EOFException` unconditionally. Let me know if I’m mistaken here!

A couple follow-ups:

I want limits overloads for `readUtf8Line()`, `indexOf(ByteString)`, and `indexOfElement()`. I’ll let you do these unless you’d rather not!

We might want to reinterpret the limit argument for `readUtf8Line()`. I think that as a caller I’m probably thinking in “line lengths” and not “delimiter offsets”. For example I might assume 72 bytes per line for an xterm. Or [RFC 2822](https://tools.ietf.org/html/rfc2822#section-3.5) specifies “maximum 998 characters” which came from 1000 characters including CRLF in an earlier spec. It’s nice if I say `readUtf8LineStrict(2048)` and I know the result fits in 2048 UTF-8 bytes. Unfortunately this has one sticky special case. If we make `readUtf8LineStrict(10)` delegate to `indexOf(0, 11)` then we might get `\r` as the byte at 10. So sometimes we’ll need to peek ahead (N+1) bytes and other times (N+2) bytes. Ick.

swankjesse commented:
PR as merged: 
https://github.com/square/okio/commit/d1d1c7bbba6561860d0e66af0900ca2ab105d814
